scripts/scripts.js

```diff
@@ -132,15 +133,228 @@ export function decorateMain(main) {
+    // Iterate over the array of items in each entry
+    if (entry.items) {
+      entry.items.forEach((item) => {
+        if (item.schema === 'https://ns.adobe.com/personalization/dom-action') {
+          const cssSelector = item.data.selector;
+          const { prehidingSelector } = item.data;
+          let section;
+          if (prehidingSelector) {
+            section = getSectionByElementSelector(prehidingSelector);
+          }
+          if (!section && cssSelector) {
+            section = getSectionByElementSelector(cssSelector);
+          }
+          if (section) {
+            section.classList.add('visibility-hidden');
+          }
+        }
+      });
+    }
+  });
+  document.body.classList.remove('visibility-hidden');
+  onDecoratedElement(async () => {
+    await window.alloy('applyPropositions', { propositions });
+
+    // keep track of propositions that were applied
+    propositions.forEach((p) => {
+      p.items = p.items.filter((i) => i.schema !== 'https://ns.adobe.com/personalization/dom-action' || !getElementForProposition(i));
+    });
+  });
+
+  // Reporting is deferred to avoid long tasks
+  window.setTimeout(() => {
+    // Report shown decisions
+    window.alloy('sendEvent', {
+      xdm: {
+        eventType: 'decisioning.propositionDisplay',
+        _experience: {
+          decisioning: { propositions },
+        },
+      },
+    });
+  });
+}
+
+async function sendAnalyticsEvent(xdmData) {
+  // eslint-disable-next-line no-undef
+  if (!alloy) {
+    return Promise.resolve();
+  }
+  // eslint-disable-next-line no-undef
+  return alloy('sendEvent', {
+    documentUnloading: true,
+    xdm: xdmData,
+  });
+}
+
+/**
+ * Basic tracking for page views with alloy
+ * @param document
+ * @param additionalXdmFields
+ * @returns {Promise<*>}
+ */
+// eslint-disable-next-line no-unused-vars
+async function analyticsTrackPageViews(document, additionalXdmFields = {}) {
+  const xdmData = {
+    eventType: 'web.webpagedetails.pageViews',
+    web: {
+      webPageDetails: {
+        pageViews: {
+          value: 1,
+        },
+        name: document.title,
+      },
+    },
+  };
+  sendAnalyticsEvent(xdmData);
+}
+
+function initializeTargetAnalytics() {
+  const alloyLoadedPromise = initWebSDK(`${window.hlx.codeBasePath}/scripts/alloy.min.js`);
+
+  if (getMetadata('target')) {
+    alloyLoadedPromise.then(() => {
+      fetch(`${getHostUrl()}/websdkconfig.json`)
+        .then((response) => response.json())
+        .then((configData) => {
+          // eslint-disable-next-line no-undef
+          alloy('configure', {
+            datastreamId: configData.data[0].DatastreamId,
+            orgId: configData.data[0].OrgId,
+          });
+          getAndApplyRenderDecisions();
+          analyticsTrackPageViews(document);
+        })
+        .catch((error) => {
+          // eslint-disable-next-line no-console
+          console.error('Error:', error);
+        });
+    });
+  }
+}
```

Comment:
I'd probably await this… otherwise you risk having the decoration happening in parallel to loading the target stuff.
```suggestion
async function initializeTargetAnalytics() {
  const alloyLoadedPromise = initWebSDK(`${window.hlx.codeBasePath}/scripts/alloy.min.js`);

  if (getMetadata('target')) {
    alloyLoadedPromise
      .then(() => fetch(`${getHostUrl()}/websdkconfig.json`))
      .then((response) => response.json())
      .then((configData) => {
        window.alloy('configure', {
          datastreamId: configData.data[0].DatastreamId,
          orgId: configData.data[0].OrgId,
        });
        await getAndApplyRenderDecisions();
        await analyticsTrackPageViews(document);
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.error('Error:', error);
      });
  }
  return alloyLoadedPromise;
}
```

Reply:
> I'd probably await this… otherwise you risk having the decoration happening in parallel to loading the target stuff.

But we are making those section visibility hidden and making them visible only when proposition is applied using await here https://github.com/aemsites/icicidirect/pull/280/files#diff-fbceb58f32b99b8a5309251f8c49ac8eda82fba58b905c600c92bce1ed2a1c8cR238
Do we still have the risk of flicker ?

Reply:
Essentially the moment the block or section is switched to visible (via the respective status attribute change), the browser gets the signal to repaint those areas. The way we fine-tuned the patching via the `MutationObserver`, essentially applies the DOM modifications at the same time, and in our experience so far, those patching operations took <20ms. Since the paint refresh rate essentially triggers only every 16ms in the browser, it would be completely seamless in most cases, so you shouldn't have to worry about flicker.

There are a few edge cases, so it's not a 100% guarantee, and if you have some we can spend some time optimizing accordingly. But as a general rule, I'd refrain from any additional show/hide logic. The boilerplate + our `MutationObserver` should have you covered out-of-the-box

Reply:
You want to make sure that the websdk is loaded as part of the eager phase so that you don't delay the personalization of the page (especially for LCP elements), otherwise you'll have flicker and CLS.
This comes with a slight trade-off on the network requests indeed.

If you want to minimize the network delay you can load both alloy and websdkconfig.json in parallel (using Promises.all) instead of chaining with a thenable:
```js
async function initializeTargetAnalytics() {
  const alloyLoadedPromise = Promise.all([
    initWebSDK(`${window.hlx.codeBasePath}/scripts/alloy.min.js`),
    fetch(`${getHostUrl()}/websdkconfig.json`).then(() => response.json()),
  ]);
  if (getMetadata('target')) {
    alloyLoadedPromise
      .then(([,configData]) => {
        window.alloy('configure', {
          datastreamId: configData.data[0].DatastreamId,
          orgId: configData.data[0].OrgId,
        });
        await getAndApplyRenderDecisions();
        await analyticsTrackPageViews(document);
      })
      .catch((error) => {
        // eslint-disable-next-line no-console
        console.error('Error:', error);
      });
  }
  return alloyLoadedPromise;
}
```

I'm going to guess that `websdkconfig.json` should also be loaded on all pages, not just the ones with target.
Theoretically, you could even inject preload HTTP headers for those 2 files (+lib-franking, scripts.js and styles.css) so they are downloaded even before the HTML `<head>` is loaded and parsed to speed things up even more. But only do it if you see perf issues

Reply:
@ramboz 
https://googlechrome.github.io/lighthouse/scorecalc/#FCP=1950.9211419506385&TTI=4319.136947009884&SI=1999.8217457632097&TBT=231.8&LCP=3243.1771451081063&CLS=0.1355305475876425&device=mobile&version=10

Best Performance I could get from this is 83.
Let me try inject preload HTTP headers for those 2 files and see if it improves the performance a bit more

Reply:
With all the martech that is loaded on the page, it's hard to know for sure what kills the score… but the top contenders I have on my end from https://pagespeed.web.dev/analysis/https-target-alloy-1--icicidirect--aemsites-hlx-live-research-equity/9iuo4yb912?form_factor=mobile are:
- google recaptcha
- google tag manager
- adobe launch

None of those are related to Target and the code you touched, really. 

Maybe try introducing a `?martech=off` flag to disable all of those and give you a more stable baseline to test for in the PRs

Reply:
The LHS/PSI report will count anything in the TBT as long as there isn't a 5s CPU idle phase.
If you have a clear-cut lazy/delayed with no loose promise, the TBT from the martech would not be counted in the report, but most implementations don't have that (for instance we typically do not await the footer loading).

When you have a rich martech stack, the easiest is to have a "feature flag" like I suggested above, and then update the URLs in your PR template to use `?martech=off` so you can follow project code perf regressions precisely, otherwise the variance coming from the martech stack will just hit you with 20~40 points randomly, and the <5 points you may lose from a PR are just swallowed in the noise

Reply:
@ramboz  I tried an approach for awaiting on header and footer. something like 
```
await loadHeader(doc.querySelector('header'));
  await loadFooter(doc.querySelector('footer'));
  await loadCSS(`${window.hlx.codeBasePath}/styles/lazy-styles.css`);
  await loadFonts();
```
That does provide some idle time before the delayed phase but it seems its not enough as well. 
We do have `loadMartech=off` in our code base which exclude launch and gtm scripts. let me try to improvise to exclude all martech to see the performance gap

Reply:
Ah I missed the `loadMartech=off`. Let me run a perf audit with it to see

Reply:
Seems all the TBT you have is coming from the recaptcha module:
<img width="951" alt="Screenshot 2024-06-05 at 10 06 14" src="https://github.com/aemsites/icicidirect/assets/1235810/63fe3a60-25c1-4b0f-aa8d-6636581411cc">

There is a bit of CLS that could be fixed by specifying image sizes in the CSS, and potentially a bit of LCP you can shave off by optimizing the image sizes to the viewport. Other than that, I don't see any noticeable impact of the Target and Analytics integration